diff --git a/contracts/suilend/sources/lending_market.move b/contracts/suilend/sources/lending_market.move
index c90807d..7e70a7a 100644
--- a/contracts/suilend/sources/lending_market.move
+++ b/contracts/suilend/sources/lending_market.move
@@ -16,6 +16,7 @@ module suilend::lending_market {
     use suilend::rate_limiter::{Self, RateLimiter, RateLimiterConfig};
     use suilend::reserve::{Self, Reserve, CToken, LiquidityRequest};
     use suilend::reserve_config::{ReserveConfig, borrow_fee};
+    use suilend::obligation::ExistStaleOracles;
 
     // === Errors ===
     const EIncorrectVersion: u64 = 1;
@@ -265,6 +266,18 @@ module suilend::lending_market {
         cap
     }
 
+    public fun refresh_obligation_health<P>(
+        lending_market: &mut LendingMarket<P>,
+        id: ID,
+        clock: &Clock
+    ) {
+        let obligation = object_table::borrow_mut(
+            &mut lending_market.obligations,
+            id,
+        );
+        obligation::refresh_health<P>(obligation, &mut lending_market.reserves, clock);
+    }
+
     /// Deposits liquidity into a reserve and mints cTokens in return.
     ///
     /// This function allows a user to deposit a certain amount of a token into a reserve
@@ -1236,6 +1249,11 @@ module suilend::lending_market {
         &lending_market.reserves
     }
 
+    #[test_only]
+    public fun reserves_mut<P>(lending_market: &mut LendingMarket<P>): &mut vector<Reserve<P>> {
+        &mut lending_market.reserves
+    }
+
     fun max_borrow_amount<P>(
         mut rate_limiter: RateLimiter,
         obligation: &Obligation<P>,
@@ -1351,6 +1369,11 @@ module suilend::lending_market {
         object_table::borrow(&lending_market.obligations, obligation_id)
     }
 
+     /// Get a reference to an obligation by its ID.
+    public fun obligation_mut<P>(lending_market: &mut LendingMarket<P>, obligation_id: ID): &mut Obligation<P> {
+        object_table::borrow_mut(&mut lending_market.obligations, obligation_id)
+    }
+
     /// Get the fee receiver address.
     public fun fee_receiver<P>(lending_market: &LendingMarket<P>): address {
         lending_market.fee_receiver
