diff --git a/contracts/suilend/sources/obligation.move b/contracts/suilend/sources/obligation.move
index df13b62..10133f4 100644
--- a/contracts/suilend/sources/obligation.move
+++ b/contracts/suilend/sources/obligation.move
@@ -155,7 +155,7 @@ module suilend::obligation {
     /// # Returns
     ///
     /// * `Obligation<P>` - A new `Obligation` instance.
-    public(package) fun create_obligation<P>(
+    public fun create_obligation<P>(
         lending_market_id: ID,
         ctx: &mut TxContext,
     ): Obligation<P> {
@@ -202,7 +202,7 @@ module suilend::obligation {
     ///
     /// * If any `reserve_array_index` in the obligation's deposits or borrows is out
     /// of bounds for the `reserves` vector.
-    public(package) fun refresh<P>(
+    public fun refresh<P>(
         obligation: &mut Obligation<P>,
         reserves: &mut vector<Reserve<P>>,
         clock: &Clock,
@@ -349,7 +349,7 @@ module suilend::obligation {
     /// * If the obligation already has a borrow for the same asset as the deposit (`ECannotDepositAndBorrowSameAsset`).
     /// * If the `reserve_array_index` of the deposit is invalid or out of bounds for the `reserves` vector.
     /// * If the `user_reward_manager_index` is invalid or out of bounds for the `user_reward_managers` vector.
-    public(package) fun deposit<P>(
+    public fun deposit<P>(
         obligation: &mut Obligation<P>,
         reserve: &mut Reserve<P>,
         clock: &Clock,
@@ -435,7 +435,7 @@ module suilend::obligation {
     /// non-isolated asset when an isolated asset is already borrowed (`EIsolatedAssetViolation`).
     /// * If the `reserve_array_index` of the borrow is invalid or out of bounds for the `reserves` vector.
     /// * If the `user_reward_manager_index` is invalid or out of bounds for the `user_reward_managers` vector.
-    public(package) fun borrow<P>(
+    public fun borrow<P>(
         obligation: &mut Obligation<P>,
         reserve: &mut Reserve<P>,
         clock: &Clock,
@@ -526,7 +526,7 @@ module suilend::obligation {
     /// for the `reserves` vector.
     /// * If the `user_reward_manager_index` is invalid or out of bounds for the
     /// `user_reward_managers` vector.
-    public(package) fun repay<P>(
+    public fun repay<P>(
         obligation: &mut Obligation<P>,
         reserve: &mut Reserve<P>,
         clock: &Clock,
@@ -647,7 +647,7 @@ module suilend::obligation {
     /// * If the obligation is not healthy after the withdrawal (`EObligationIsNotHealthy`).
     /// * If the `reserve_array_index` of the deposit is invalid or out of bounds for the `reserves` vector.
     /// * If the `user_reward_manager_index` is invalid or out of bounds for the `user_reward_managers` vector.
-    public(package) fun withdraw<P>(
+    public fun withdraw<P>(
         obligation: &mut Obligation<P>,
         reserve: &mut Reserve<P>,
         clock: &Clock,
@@ -699,7 +699,7 @@ module suilend::obligation {
     /// * If the `repay_reserve_array_index` or `withdraw_reserve_array_index` are out of bounds for the `reserves` vector.
     /// * If the obligation does not have a borrow for the specified repay reserve.
     /// * If the obligation does not have a deposit for the specified withdraw reserve.
-    public(package) fun liquidate<P>(
+    public fun liquidate<P>(
         obligation: &mut Obligation<P>,
         reserves: &mut vector<Reserve<P>>,
         repay_reserve_array_index: u64,
@@ -810,7 +810,7 @@ module suilend::obligation {
     /// This function will panic under the following conditions:
     /// * If the obligation is not forgivable (i.e., it still has deposits).
     /// * If the obligation does not have a borrow for the specified reserve.
-    public(package) fun forgive<P>(
+    public fun forgive<P>(
         obligation: &mut Obligation<P>,
         reserve: &mut Reserve<P>,
         clock: &Clock,
@@ -851,7 +851,7 @@ module suilend::obligation {
     ///   in the obligation.
     /// * If the `reward_index` is out of bounds for the rewards in the `pool_reward_manager`.
     /// * It will also propagate any panics from `liquidity_mining::claim_rewards`.
-    public(package) fun claim_rewards<P, T>(
+    public fun claim_rewards<P, T>(
         obligation: &mut Obligation<P>,
         pool_reward_manager: &mut PoolRewardManager,
         clock: &Clock,
@@ -1067,7 +1067,7 @@ module suilend::obligation {
     /// # Panics
     ///
     /// * If the `borrow_weight` of the reserve is zero, as this would lead to division by zero.
-    public(package) fun max_borrow_amount<P>(
+    public fun max_borrow_amount<P>(
         obligation: &Obligation<P>,
         reserve: &Reserve<P>,
     ): u64 {
@@ -1106,7 +1106,7 @@ module suilend::obligation {
     ///
     /// * Panics with `EDepositNotFound` if the obligation does not have a deposit for the
     ///   specified `reserve`.
-    public(package) fun max_withdraw_amount<P>(
+    public fun max_withdraw_amount<P>(
         obligation: &Obligation<P>,
         reserve: &Reserve<P>,
     ): u64 {
@@ -1159,7 +1159,7 @@ module suilend::obligation {
     /// # Panics
     ///
     /// * Panics with `EOraclesAreStale` if `exist_stale_oracles` is `Some`.
-    public(package) fun assert_no_stale_oracles(exist_stale_oracles: Option<ExistStaleOracles>) {
+    public fun assert_no_stale_oracles(exist_stale_oracles: Option<ExistStaleOracles>) {
         assert!(option::is_none(&exist_stale_oracles), EOraclesAreStale);
         option::destroy_none(exist_stale_oracles);
     }
@@ -1185,7 +1185,7 @@ module suilend::obligation {
     ///   deposit or borrow in the obligation has an invalid `reserve_array_index` or
     ///   `user_reward_manager_index` that is out of bounds for the `reserves` or
     ///   `user_reward_managers` vectors, respectively.
-    public(package) fun zero_out_rewards_if_looped<P>(
+    public fun zero_out_rewards_if_looped<P>(
         obligation: &mut Obligation<P>,
         reserves: &mut vector<Reserve<P>>,
         clock: &Clock,
@@ -1206,7 +1206,7 @@ module suilend::obligation {
     /// # Panics
     /// * If the internal `target_reserve_array_indices` and `disabled_pairings_map` vectors
     ///   have inconsistent lengths, leading to an out-of-bounds access.
-    public(package) fun is_looped<P>(obligation: &Obligation<P>): bool {
+    public fun is_looped<P>(obligation: &Obligation<P>): bool {
         let target_reserve_array_indices = vector[1, 2, 5, 7, 19, 20, 3, 9];
 
         // The vector target_reserve_array_indices maps to disabled_pairings_map
@@ -1512,14 +1512,14 @@ module suilend::obligation {
         i
     }
 
-    public(package) fun find_borrow<P>(obligation: &Obligation<P>, reserve: &Reserve<P>): &Borrow {
+    public fun find_borrow<P>(obligation: &Obligation<P>, reserve: &Reserve<P>): &Borrow {
         let i = find_borrow_index(obligation, reserve);
         assert!(i < vector::length(&obligation.borrows), EBorrowNotFound);
 
         vector::borrow(&obligation.borrows, i)
     }
 
-    public(package) fun find_deposit<P>(
+    public fun find_deposit<P>(
         obligation: &Obligation<P>,
         reserve: &Reserve<P>,
     ): &Deposit {
@@ -1587,7 +1587,7 @@ module suilend::obligation {
         vector::length(&obligation.deposits) - 1
     }
 
-    public(package) fun find_user_reward_manager_index<P>(
+    public fun find_user_reward_manager_index<P>(
         obligation: &Obligation<P>,
         pool_reward_manager: &PoolRewardManager,
     ): u64 {
@@ -1637,7 +1637,7 @@ module suilend::obligation {
     }
 
     #[test_only]
-    public(package) fun borrows_mut<P>(obligation: &mut Obligation<P>): &mut vector<Borrow> {
+    public fun borrows_mut<P>(obligation: &mut Obligation<P>): &mut vector<Borrow> {
         &mut obligation.borrows
     }
 
