diff --git a/contracts/suilend/sources/obligation.move b/contracts/suilend/sources/obligation.move
index df13b62..ebc830a 100644
--- a/contracts/suilend/sources/obligation.move
+++ b/contracts/suilend/sources/obligation.move
@@ -155,7 +155,7 @@ module suilend::obligation {
     /// # Returns
     ///
     /// * `Obligation<P>` - A new `Obligation` instance.
-    public(package) fun create_obligation<P>(
+    public fun create_obligation<P>(
         lending_market_id: ID,
         ctx: &mut TxContext,
     ): Obligation<P> {
@@ -202,7 +202,7 @@ module suilend::obligation {
     ///
     /// * If any `reserve_array_index` in the obligation's deposits or borrows is out
     /// of bounds for the `reserves` vector.
-    public(package) fun refresh<P>(
+    public fun refresh<P>(
         obligation: &mut Obligation<P>,
         reserves: &mut vector<Reserve<P>>,
         clock: &Clock,
@@ -323,6 +323,88 @@ module suilend::obligation {
         option::none()
     }
 
+    /// Refreshes only the obligation's health-related values:
+    /// - `deposited_value_usd`
+    /// - `allowed_borrow_value_usd`
+    /// - `weighted_borrowed_value_upper_bound_usd`
+    public fun refresh_health<P>(
+        obligation: &mut Obligation<P>,
+        reserves: &mut vector<Reserve<P>>,
+        clock: &Clock,
+    ) {
+    
+        let mut i = 0;
+        let mut deposited_value_usd = decimal::from(0);
+        let mut allowed_borrow_value_usd = decimal::from(0);
+
+        while (i < vector::length(&obligation.deposits)) {
+            let deposit = &mut obligation.deposits[i];
+
+            let deposit_reserve = &mut reserves[deposit.reserve_array_index];
+
+            reserve::compound_interest(deposit_reserve, clock);
+
+            let market_value = reserve::ctoken_market_value(
+                deposit_reserve,
+                deposit.deposited_ctoken_amount,
+            );
+            let market_value_lower_bound = reserve::ctoken_market_value_lower_bound(
+                deposit_reserve,
+                deposit.deposited_ctoken_amount,
+            );
+
+            deposit.market_value = market_value;
+            deposited_value_usd = add(deposited_value_usd, market_value);
+            allowed_borrow_value_usd =
+                add(
+                    allowed_borrow_value_usd,
+                    mul(
+                        market_value_lower_bound,
+                        open_ltv(config(deposit_reserve)),
+                    ),
+                );
+
+            i = i + 1;
+        };
+
+        obligation.deposited_value_usd = deposited_value_usd;
+        obligation.allowed_borrow_value_usd = allowed_borrow_value_usd;
+
+        let mut i = 0;
+        let mut weighted_borrowed_value_upper_bound_usd = decimal::from(0);
+
+        while (i < vector::length(&obligation.borrows)) {
+            let borrow = &mut obligation.borrows[i];
+
+            let borrow_reserve = &mut reserves[borrow.reserve_array_index];
+            reserve::compound_interest(borrow_reserve, clock);
+            compound_debt(borrow, borrow_reserve);
+
+            let market_value = reserve::market_value(borrow_reserve, borrow.borrowed_amount);
+            let market_value_upper_bound = reserve::market_value_upper_bound(
+                borrow_reserve,
+                borrow.borrowed_amount,
+            );
+            let borrow_weight = borrow_weight(config(borrow_reserve));
+            borrow.market_value = market_value;
+
+            weighted_borrowed_value_upper_bound_usd =
+                add(
+                    weighted_borrowed_value_upper_bound_usd,
+                    mul(
+                        market_value_upper_bound,
+                        borrow_weight,
+                    ),
+                );
+
+            i = i + 1;
+        };
+
+        obligation.weighted_borrowed_value_upper_bound_usd =
+            weighted_borrowed_value_upper_bound_usd;
+
+    }
+
     /// Processes a deposit into the obligation.
     ///
     /// This function handles the logic for depositing a specified amount of ctokens into a
@@ -349,7 +431,7 @@ module suilend::obligation {
     /// * If the obligation already has a borrow for the same asset as the deposit (`ECannotDepositAndBorrowSameAsset`).
     /// * If the `reserve_array_index` of the deposit is invalid or out of bounds for the `reserves` vector.
     /// * If the `user_reward_manager_index` is invalid or out of bounds for the `user_reward_managers` vector.
-    public(package) fun deposit<P>(
+    public fun deposit<P>(
         obligation: &mut Obligation<P>,
         reserve: &mut Reserve<P>,
         clock: &Clock,
@@ -435,7 +517,7 @@ module suilend::obligation {
     /// non-isolated asset when an isolated asset is already borrowed (`EIsolatedAssetViolation`).
     /// * If the `reserve_array_index` of the borrow is invalid or out of bounds for the `reserves` vector.
     /// * If the `user_reward_manager_index` is invalid or out of bounds for the `user_reward_managers` vector.
-    public(package) fun borrow<P>(
+    public fun borrow<P>(
         obligation: &mut Obligation<P>,
         reserve: &mut Reserve<P>,
         clock: &Clock,
@@ -526,7 +608,7 @@ module suilend::obligation {
     /// for the `reserves` vector.
     /// * If the `user_reward_manager_index` is invalid or out of bounds for the
     /// `user_reward_managers` vector.
-    public(package) fun repay<P>(
+    public fun repay<P>(
         obligation: &mut Obligation<P>,
         reserve: &mut Reserve<P>,
         clock: &Clock,
@@ -647,7 +729,7 @@ module suilend::obligation {
     /// * If the obligation is not healthy after the withdrawal (`EObligationIsNotHealthy`).
     /// * If the `reserve_array_index` of the deposit is invalid or out of bounds for the `reserves` vector.
     /// * If the `user_reward_manager_index` is invalid or out of bounds for the `user_reward_managers` vector.
-    public(package) fun withdraw<P>(
+    public fun withdraw<P>(
         obligation: &mut Obligation<P>,
         reserve: &mut Reserve<P>,
         clock: &Clock,
@@ -699,7 +781,7 @@ module suilend::obligation {
     /// * If the `repay_reserve_array_index` or `withdraw_reserve_array_index` are out of bounds for the `reserves` vector.
     /// * If the obligation does not have a borrow for the specified repay reserve.
     /// * If the obligation does not have a deposit for the specified withdraw reserve.
-    public(package) fun liquidate<P>(
+    public fun liquidate<P>(
         obligation: &mut Obligation<P>,
         reserves: &mut vector<Reserve<P>>,
         repay_reserve_array_index: u64,
@@ -714,6 +796,33 @@ module suilend::obligation {
         let borrow = find_borrow(obligation, repay_reserve);
         let deposit = find_deposit(obligation, withdraw_reserve);
 
+        let (final_settle_amount, final_withdraw_amount) = liquidation_amounts(
+            obligation, 
+            repay_amount, 
+            withdraw_reserve,
+            repay_reserve,
+            borrow,
+            deposit);
+
+
+        repay(
+            obligation,
+            vector::borrow_mut(reserves, repay_reserve_array_index),
+            clock,
+            final_settle_amount,
+        );
+        withdraw_unchecked(
+            obligation,
+            vector::borrow_mut(reserves, withdraw_reserve_array_index),
+            clock,
+            final_withdraw_amount,
+        );
+
+        log_obligation_data(obligation);
+        (final_withdraw_amount, final_settle_amount)
+    }
+
+    public fun liquidation_amounts<P>(obligation: &Obligation<P>, repay_amount: u64, withdraw_reserve: &Reserve<P>, repay_reserve: &Reserve<P>, borrow: &Borrow, deposit: &Deposit): (Decimal, u64) {
         // invariant: repay_amount <= borrow.borrowed_amount
         let repay_amount = if (le(borrow.market_value, decimal::from(1))) {
             // full liquidation
@@ -761,30 +870,16 @@ module suilend::obligation {
             final_settle_amount = mul(repay_amount, repay_pct);
             final_withdraw_amount = deposit.deposited_ctoken_amount;
         } else {
-            let withdraw_pct = div(withdraw_value, deposit.market_value);
-
+            let withdraw_pct = div(withdraw_value, deposit.market_value); // percentage of the VALUE (USD) we withdraw
             final_settle_amount = repay_amount;
+            
+            // The amount of ctokens corresponding to that value
             final_withdraw_amount =
                 floor(
-                    mul(decimal::from(deposit.deposited_ctoken_amount), withdraw_pct),
+                    mul(decimal::from(deposit.deposited_ctoken_amount), withdraw_pct), 
                 );
         };
-
-        repay(
-            obligation,
-            vector::borrow_mut(reserves, repay_reserve_array_index),
-            clock,
-            final_settle_amount,
-        );
-        withdraw_unchecked(
-            obligation,
-            vector::borrow_mut(reserves, withdraw_reserve_array_index),
-            clock,
-            final_withdraw_amount,
-        );
-
-        log_obligation_data(obligation);
-        (final_withdraw_amount, final_settle_amount)
+        (final_settle_amount, final_withdraw_amount)
     }
 
     /// Forgives a portion of a borrow for an obligation that has no deposits.
@@ -810,7 +905,7 @@ module suilend::obligation {
     /// This function will panic under the following conditions:
     /// * If the obligation is not forgivable (i.e., it still has deposits).
     /// * If the obligation does not have a borrow for the specified reserve.
-    public(package) fun forgive<P>(
+    public fun forgive<P>(
         obligation: &mut Obligation<P>,
         reserve: &mut Reserve<P>,
         clock: &Clock,
@@ -851,7 +946,7 @@ module suilend::obligation {
     ///   in the obligation.
     /// * If the `reward_index` is out of bounds for the rewards in the `pool_reward_manager`.
     /// * It will also propagate any panics from `liquidity_mining::claim_rewards`.
-    public(package) fun claim_rewards<P, T>(
+    public fun claim_rewards<P, T>(
         obligation: &mut Obligation<P>,
         pool_reward_manager: &mut PoolRewardManager,
         clock: &Clock,
@@ -891,31 +986,62 @@ module suilend::obligation {
         obligation.deposited_value_usd
     }
 
+    /// Get the deposited value in USD.
+    public fun deposited_value_usd_mut<P>(obligation: &mut Obligation<P>): &mut Decimal {
+        &mut obligation.deposited_value_usd
+    }
+
     /// Get the allowed borrow value in USD.
     public fun allowed_borrow_value_usd<P>(obligation: &Obligation<P>): Decimal {
         obligation.allowed_borrow_value_usd
     }
 
+    /// Get the allowed borrow value in USD.
+    public fun allowed_borrow_value_usd_mut<P>(obligation: &mut Obligation<P>): &mut Decimal {
+        &mut obligation.allowed_borrow_value_usd
+    }
+
     /// Get the unhealthy borrow value in USD.
     public fun unhealthy_borrow_value_usd<P>(obligation: &Obligation<P>): Decimal {
         obligation.unhealthy_borrow_value_usd
     }
 
+    /// Get the unhealthy borrow value in USD.
+    public fun unhealthy_borrow_value_usd_mut<P>(obligation: &mut Obligation<P>): &mut Decimal {
+        &mut obligation.unhealthy_borrow_value_usd
+    }
+
     /// Get the unweighted borrowed value in USD.
     public fun unweighted_borrowed_value_usd<P>(obligation: &Obligation<P>): Decimal {
         obligation.unweighted_borrowed_value_usd
     }
 
+    /// Get the unweighted borrowed value in USD.
+    public fun unweighted_borrowed_value_usd_mut<P>(obligation: &mut Obligation<P>): &mut Decimal {
+        &mut obligation.unweighted_borrowed_value_usd
+    }
+
+
     /// Get the weighted borrowed value in USD.
     public fun weighted_borrowed_value_usd<P>(obligation: &Obligation<P>): Decimal {
         obligation.weighted_borrowed_value_usd
     }
 
+    /// Get the weighted borrowed value in USD.
+    public fun weighted_borrowed_value_usd_mut<P>(obligation: &mut Obligation<P>): &mut Decimal {
+        &mut obligation.weighted_borrowed_value_usd
+    }
+
     /// Get the weighted borrowed value upper bound in USD.
     public fun weighted_borrowed_value_upper_bound_usd<P>(obligation: &Obligation<P>): Decimal {
         obligation.weighted_borrowed_value_upper_bound_usd
     }
 
+    /// Get the weighted borrowed value upper bound in USD.
+    public fun weighted_borrowed_value_upper_bound_usd_mut<P>(obligation: &mut Obligation<P>): &mut Decimal {
+        &mut obligation.weighted_borrowed_value_upper_bound_usd
+    }
+
     /// Whether the obligation is borrowing an isolated asset.
     public fun borrowing_isolated_asset<P>(obligation: &Obligation<P>): bool {
         obligation.borrowing_isolated_asset
@@ -926,6 +1052,10 @@ module suilend::obligation {
         &obligation.user_reward_managers
     }
 
+    public fun user_reward_managers_mut<P>(obligation: &mut Obligation<P>): &mut vector<UserRewardManager> {
+        &mut obligation.user_reward_managers
+    }
+
     public use fun deposit_coin_type as Deposit.coin_type;
 
     /// Get the coin type of the deposit.
@@ -947,6 +1077,13 @@ module suilend::obligation {
         deposit.deposited_ctoken_amount
     }
 
+    public use fun deposit_deposited_ctoken_amount_mut as Deposit.deposited_ctoken_amount_mut;
+
+    /// Get the deposited ctoken amount of the deposit.
+    public fun deposit_deposited_ctoken_amount_mut(deposit: &mut Deposit): &mut u64 {
+        &mut deposit.deposited_ctoken_amount
+    }
+
     public use fun deposit_market_value as Deposit.market_value;
 
     /// Get the market value of the deposit.
@@ -954,6 +1091,13 @@ module suilend::obligation {
         deposit.market_value
     }
 
+    public use fun deposit_market_value_mut as Deposit.market_value_mut;
+
+    /// Get the market value of the deposit.
+    public fun deposit_market_value_mut(deposit: &mut Deposit): &mut Decimal {
+        &mut deposit.market_value
+    }
+
     public use fun deposit_user_reward_manager_index as Deposit.user_reward_manager_index;
 
     /// Get the user reward manager index of the deposit.
@@ -982,6 +1126,14 @@ module suilend::obligation {
         borrow.borrowed_amount
     }
 
+    public use fun borrow_borrowed_amount_mut as Borrow.borrowed_amount_mut;
+
+    /// Get the borrowed amount of the borrow.
+    public fun borrow_borrowed_amount_mut(borrow: &mut Borrow): &mut Decimal {
+        &mut borrow.borrowed_amount
+    }
+
+
     public use fun borrow_cumulative_borrow_rate as Borrow.cumulative_borrow_rate;
 
     /// Get the cumulative borrow rate of the borrow.
@@ -996,6 +1148,13 @@ module suilend::obligation {
         borrow.market_value
     }
 
+    public use fun borrow_market_value_mut as Borrow.market_value_mut;
+
+    /// Get the market value of the borrow.
+    public fun borrow_market_value_mut(borrow: &mut Borrow): &mut Decimal {
+        &mut borrow.market_value
+    }
+
     public use fun borrow_user_reward_manager_index as Borrow.user_reward_manager_index;
 
     /// Get the user reward manager index of the borrow.
@@ -1018,6 +1177,18 @@ module suilend::obligation {
         0
     }
 
+    
+    public fun total_deposited_ctokens<P>(obligation: &Obligation<P>): u64 {
+        let mut v = 0;
+        let mut i = 0;
+        while (i < vector::length(&obligation.deposits)) {
+            let deposit = vector::borrow(&obligation.deposits, i);
+            v = v + deposit.deposited_ctoken_amount;
+            i = i + 1;
+        };
+        v
+    }
+
     /// Get the borrowed amount of a specific coin type.
     public fun borrowed_amount<P, T>(obligation: &Obligation<P>): Decimal {
         let mut i = 0;
@@ -1033,6 +1204,17 @@ module suilend::obligation {
         decimal::from(0)
     }
 
+    public fun total_borrowed_amount<P>(obligation: &Obligation<P>): Decimal {
+        let mut v = decimal::from(0);
+        let mut i = 0;
+        while (i < vector::length(&obligation.borrows)) {
+            let borrow = vector::borrow(&obligation.borrows, i);
+            v = v.add(borrow.borrowed_amount);
+            i = i + 1;
+        };
+        v
+    }
+
     /// Whether the obligation is healthy.
     public fun is_healthy<P>(obligation: &Obligation<P>): bool {
         le(obligation.weighted_borrowed_value_upper_bound_usd, obligation.allowed_borrow_value_usd)
@@ -1067,7 +1249,7 @@ module suilend::obligation {
     /// # Panics
     ///
     /// * If the `borrow_weight` of the reserve is zero, as this would lead to division by zero.
-    public(package) fun max_borrow_amount<P>(
+    public fun max_borrow_amount<P>(
         obligation: &Obligation<P>,
         reserve: &Reserve<P>,
     ): u64 {
@@ -1106,7 +1288,7 @@ module suilend::obligation {
     ///
     /// * Panics with `EDepositNotFound` if the obligation does not have a deposit for the
     ///   specified `reserve`.
-    public(package) fun max_withdraw_amount<P>(
+    public fun max_withdraw_amount<P>(
         obligation: &Obligation<P>,
         reserve: &Reserve<P>,
     ): u64 {
@@ -1159,11 +1341,15 @@ module suilend::obligation {
     /// # Panics
     ///
     /// * Panics with `EOraclesAreStale` if `exist_stale_oracles` is `Some`.
-    public(package) fun assert_no_stale_oracles(exist_stale_oracles: Option<ExistStaleOracles>) {
+    public fun assert_no_stale_oracles(exist_stale_oracles: Option<ExistStaleOracles>) {
         assert!(option::is_none(&exist_stale_oracles), EOraclesAreStale);
         option::destroy_none(exist_stale_oracles);
     }
 
+    public fun id<P>(obligation: &Obligation<P>): &UID {
+        &obligation.id
+    }
+
     /// Checks if an obligation is in a "looped" state and zeroes out its liquidity mining
     /// rewards if it is.
     ///
@@ -1185,7 +1371,7 @@ module suilend::obligation {
     ///   deposit or borrow in the obligation has an invalid `reserve_array_index` or
     ///   `user_reward_manager_index` that is out of bounds for the `reserves` or
     ///   `user_reward_managers` vectors, respectively.
-    public(package) fun zero_out_rewards_if_looped<P>(
+    public fun zero_out_rewards_if_looped<P>(
         obligation: &mut Obligation<P>,
         reserves: &mut vector<Reserve<P>>,
         clock: &Clock,
@@ -1206,7 +1392,7 @@ module suilend::obligation {
     /// # Panics
     /// * If the internal `target_reserve_array_indices` and `disabled_pairings_map` vectors
     ///   have inconsistent lengths, leading to an out-of-bounds access.
-    public(package) fun is_looped<P>(obligation: &Obligation<P>): bool {
+    public fun is_looped<P>(obligation: &Obligation<P>): bool {
         let target_reserve_array_indices = vector[1, 2, 5, 7, 19, 20, 3, 9];
 
         // The vector target_reserve_array_indices maps to disabled_pairings_map
@@ -1450,7 +1636,7 @@ module suilend::obligation {
     }
 
     /// Compound the debt on a borrow object
-    fun compound_debt<P>(borrow: &mut Borrow, reserve: &Reserve<P>) {
+    public fun compound_debt<P>(borrow: &mut Borrow, reserve: &Reserve<P>) {
         let new_cumulative_borrow_rate = reserve::cumulative_borrow_rate(reserve);
 
         let compounded_interest_rate = div(
@@ -1467,7 +1653,7 @@ module suilend::obligation {
         borrow.cumulative_borrow_rate = new_cumulative_borrow_rate;
     }
 
-    fun find_deposit_index<P>(obligation: &Obligation<P>, reserve: &Reserve<P>): u64 {
+    public fun find_deposit_index<P>(obligation: &Obligation<P>, reserve: &Reserve<P>): u64 {
         let mut i = 0;
         while (i < vector::length(&obligation.deposits)) {
             let deposit = vector::borrow(&obligation.deposits, i);
@@ -1498,7 +1684,7 @@ module suilend::obligation {
         i
     }
 
-    fun find_borrow_index<P>(obligation: &Obligation<P>, reserve: &Reserve<P>): u64 {
+    public fun find_borrow_index<P>(obligation: &Obligation<P>, reserve: &Reserve<P>): u64 {
         let mut i = 0;
         while (i < vector::length(&obligation.borrows)) {
             let borrow = vector::borrow(&obligation.borrows, i);
@@ -1512,14 +1698,21 @@ module suilend::obligation {
         i
     }
 
-    public(package) fun find_borrow<P>(obligation: &Obligation<P>, reserve: &Reserve<P>): &Borrow {
+    public fun find_borrow<P>(obligation: &Obligation<P>, reserve: &Reserve<P>): &Borrow {
         let i = find_borrow_index(obligation, reserve);
         assert!(i < vector::length(&obligation.borrows), EBorrowNotFound);
 
         vector::borrow(&obligation.borrows, i)
     }
 
-    public(package) fun find_deposit<P>(
+    public fun find_borrow_mut<P>(obligation: &mut Obligation<P>, reserve: &Reserve<P>): &mut Borrow {
+        let i = find_borrow_index(obligation, reserve);
+        assert!(i < vector::length(&obligation.borrows), EBorrowNotFound);
+
+        vector::borrow_mut(&mut obligation.borrows, i)
+    }
+
+    public fun find_deposit<P>(
         obligation: &Obligation<P>,
         reserve: &Reserve<P>,
     ): &Deposit {
@@ -1529,7 +1722,7 @@ module suilend::obligation {
         vector::borrow(&obligation.deposits, i)
     }
 
-    fun find_or_add_borrow<P>(
+    public fun find_or_add_borrow<P>(
         obligation: &mut Obligation<P>,
         reserve: &mut Reserve<P>,
         clock: &Clock,
@@ -1558,7 +1751,7 @@ module suilend::obligation {
         vector::length(&obligation.borrows) - 1
     }
 
-    fun find_or_add_deposit<P>(
+    public fun find_or_add_deposit<P>(
         obligation: &mut Obligation<P>,
         reserve: &mut Reserve<P>,
         clock: &Clock,
@@ -1587,7 +1780,7 @@ module suilend::obligation {
         vector::length(&obligation.deposits) - 1
     }
 
-    public(package) fun find_user_reward_manager_index<P>(
+    public fun find_user_reward_manager_index<P>(
         obligation: &Obligation<P>,
         pool_reward_manager: &PoolRewardManager,
     ): u64 {
@@ -1637,7 +1830,7 @@ module suilend::obligation {
     }
 
     #[test_only]
-    public(package) fun borrows_mut<P>(obligation: &mut Obligation<P>): &mut vector<Borrow> {
+    public fun borrows_mut<P>(obligation: &mut Obligation<P>): &mut vector<Borrow> {
         &mut obligation.borrows
     }
 
